### PR TITLE
Security Basic

### PR DESCRIPTION
**Pull request by katrevichdv**

### CODE REVIEW

File: src/main/java/com/example/bookstore/controller/AuthenticationController.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+import com.example.bookstore.exception.RegistrationException;
+import com.example.bookstore.service.UserService;
+import jakarta.validation.Valid;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping(value = "/api/auth")
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookstore/controller/AuthenticationController.java]:
```suggestion
@RequestMapping("/auth")
```

add `server.servlet.context-path=/api` to application.properties

File: src/main/java/com/example/bookstore/dto/user/UserRequestDto.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+import com.example.bookstore.validation.constraints.FieldMatch;
+import jakarta.validation.constraints.Email;
+import jakarta.validation.constraints.NotBlank;
+import jakarta.validation.constraints.NotNull;
+import jakarta.validation.constraints.Size;
+
+@FieldMatch(field = "password", fieldMatch = "repeatPassword", message = "Passwords don't match")
+public record UserRequestDto(
+        @NotNull
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookstore/dto/user/UserRequestDto.java]:
```suggestion
        @NotBlank
```

File: src/main/java/com/example/bookstore/model/User.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
+package com.example.bookstore.model;
+
+import jakarta.persistence.Column;
+import jakarta.persistence.Entity;
+import jakarta.persistence.GeneratedValue;
+import jakarta.persistence.GenerationType;
+import jakarta.persistence.Id;
+import lombok.Data;
+
+@Data
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookstore/model/User.java]: Don't use `@Data` for entities

File: src/main/java/com/example/bookstore/service/impl/UserServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+import com.example.bookstore.dto.user.UserRequestDto;
+import com.example.bookstore.dto.user.UserResponseDto;
+import com.example.bookstore.exception.RegistrationException;
+import com.example.bookstore.mapper.UserMapper;
+import com.example.bookstore.repository.user.UserRepository;
+import com.example.bookstore.service.UserService;
+import org.springframework.stereotype.Service;
+
+@Service
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookstore/service/impl/UserServiceImpl.java:13]:
```suggestion
@Service
@RequiredArgsConstructor
```

File: src/main/java/com/example/bookstore/service/impl/UserServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+public class UserServiceImpl implements UserService {
+    private final UserMapper userMapper;
+    private final UserRepository userRepository;
+
+    public UserServiceImpl(UserMapper userMapper, UserRepository userRepository) {
+        this.userMapper = userMapper;
+        this.userRepository = userRepository;
+    }
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookstore/service/impl/UserServiceImpl.java]:
```suggestion
```

File: src/main/java/com/example/bookstore/service/impl/UserServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+    @Override
+    public UserResponseDto register(UserRequestDto requestDto) throws RegistrationException {
+        if (userRepository.existsByEmail(requestDto.email())) {
+            throw new RegistrationException("Can't register user. Email is taken");
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookstore/service/impl/UserServiceImpl.java]: add email to message
It's better: User with this email: already exists

File: src/main/java/com/example/bookstore/validation/validators/FieldMatchValidator.java [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+    private String fieldMatch;
+
+    @Override
+    public void initialize(FieldMatch constraintAnnotation) {
+        this.field = constraintAnnotation.field();
+        this.fieldMatch = constraintAnnotation.fieldMatch();
+    }
+
+    @Override
+    public boolean isValid(Object o, ConstraintValidatorContext constraintValidatorContext) {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookstore/validation/validators/FieldMatchValidator.java]: What is it "o" ?

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Good job! Let's improve your solution

File: src/main/java/com/example/bookstore/dto/user/UserRequestDto.java (hunk 2 above)
Comment (katrevichdv) [reply src/main/java/com/example/bookstore/dto/user/UserRequestDto.java:11]: What's the point?

File: src/main/java/com/example/bookstore/service/impl/UserServiceImpl.java (hunk 6 above)
Comment (katrevichdv) [reply src/main/java/com/example/bookstore/service/impl/UserServiceImpl.java]: I don't feel like adding users credentials to the error messages.

File: src/main/java/com/example/bookstore/model/User.java [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+import jakarta.persistence.Entity;
+import jakarta.persistence.GeneratedValue;
+import jakarta.persistence.GenerationType;
+import jakarta.persistence.Id;
+import lombok.Getter;
+import lombok.Setter;
+
+@Getter
+@Setter
+@Entity(name = "users")
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookstore/model/User.java]:
```suggestion
@Entity
@Table.....
```

File: src/main/java/com/example/bookstore/service/impl/UserServiceImpl.java (hunk 6 above)
Comment (Elena-Bruyako) [reply src/main/java/com/example/bookstore/service/impl/UserServiceImpl.java]: I didn't mention password that should be secured

File: src/main/java/com/example/bookstore/service/impl/UserServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+public class UserServiceImpl implements UserService {
+    private final UserMapper userMapper;
+    private final UserRepository userRepository;
+
+    @Override
+    public UserResponseDto register(UserRequestDto requestDto) throws RegistrationException {
+        if (userRepository.existsByEmail(requestDto.email())) {
+            throw new RegistrationException("Can't register user. Email is taken");
+        }
+        return userMapper.toDto(userRepository.save(userMapper.toModel(requestDto)));
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookstore/service/impl/UserServiceImpl.java]: It's difficult to read and debug such in line code

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see pr

File: src/main/java/com/example/bookstore/service/impl/UserServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+@Service
+@RequiredArgsConstructor
+public class UserServiceImpl implements UserService {
+    private final UserMapper userMapper;
+    private final UserRepository userRepository;
+
+    @Override
+    public UserResponseDto register(UserRequestDto requestDto) throws RegistrationException {
+        if (userRepository.existsByEmail(requestDto.email())) {
+            throw RegistrationException.emailExist(requestDto.email());
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/bookstore/service/impl/UserServiceImpl.java]: don't create a static method in the exception class, you won't need it anymore. it's not a class property and looks weird

**Review by liliia-ponomarenko — APPROVED** [review]
Good job! But please fix my comment before merging ;)

**Comment by katrevichdv** [comment]
Added User entity
Added the required DTOs and mapper
Added AuthenticationController
Added data validation
Added FieldMatch annotation

**Pull request merged**